### PR TITLE
Tensor expr fuser pass for extremely simple expressions

### PR DESCRIPTION
**Pull request by bertmaher**
Apologies for the workflow churn, but you'll need to `cmake -DBUILD_TX_TESTS=ON` to keep using our old workflow.

### CODE REVIEW

**Comment by resistor** [comment]
I think you need to rebase this to make a sane PR?

**Comment by bertmaher** [comment]
urk, I guess when MikeZ force pushed the new rebased branch it caused this PR to become horrible.  In any case it's not really a sane PR even when correctly rebased, just a WIP in case y'all had comments.  I should have something working e2e in a little bit that I can push as a semi-real thing.

File: torch/csrc/jit/passes/tensorexpr_fuser.cpp [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+      });
+  return result;
+}
+
+bool isSupported(Node* node) {
+  // TODO:
+  return node->kind() == Symbol::fromQualString("aten::add");
+}
+
+bool canHandle(Node* node, AliasDb& aliasDb) {
```
Comment (resistor) [inline torch/csrc/jit/passes/tensorexpr_fuser.cpp:49]: Probably want a check for immutability here?

Comment (resistor) [reply torch/csrc/jit/passes/tensorexpr_fuser.cpp:49]: Nvm, I see that you're handling it below.

File: torch/csrc/jit/passes/tensorexpr_fuser.cpp [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+    if (i->node()->owningBlock() == block) {
+      result.push_back(i);
+    }
+  }
+  // Sort in reverse topological order
+  std::sort(
+      result.begin(),
+      result.end(),
+      [&](torch::jit::Value* a, torch::jit::Value* b) {
+        return a->node()->isAfter(b->node());
```
Comment (zheng-xq) [inline torch/csrc/jit/passes/tensorexpr_fuser.cpp:39]: Bert and I chatted offline. 

1. isAfter is not a O(1) operation. 
2. std::sort does not work very well for partial orders. Reverse post-order DFS is a much better alternative.

File: torch/csrc/jit/passes/tensorexpr_fuser.cpp [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+  if (node->kind() == prim::Constant) {
+    return true;
+  }
+  if (node->kind() == prim::Loop) {
+    return false; // TODO
+  }
+  return isSupported(node);
+}
+
+#define REQ(cond)                           \
```
Comment (zheng-xq) [inline torch/csrc/jit/passes/tensorexpr_fuser.cpp:59]: Maybe a MERGE_REQUIRES with full name? Also undef it when you are done. 

Minor nitpicking: in this particular case, I actually find it slightly easier to read with all the code inlined without the macro.

File: torch/csrc/jit/passes/tensorexpr_fuser.cpp [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+          auto tt = n->output()->type()->cast<TensorType>();
+          auto exprDims = texprSizes(tt->sizes());
+          std::vector<DimArg> dims(exprDims.begin(), exprDims.end());
+
+          auto lt = buffers.at(luniq);
+          auto rt = buffers.at(runiq);
+          tensors.emplace(
+              n->output()->unique(),
+              Compute(
+                  "aten__add", dims, [lt, rt](const std::vector<Var>& axes) {
```
Comment (zheng-xq) [inline torch/csrc/jit/passes/tensorexpr_fuser.cpp:266]: "aten__add" -> "aten_add"?

File: torch/csrc/jit/passes/tensorexpr_fuser.cpp [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+                    return lt(axes[0]) + rt(axes[0]);
```
Comment (zheng-xq) [inline torch/csrc/jit/passes/tensorexpr_fuser.cpp:267]: If you only want to use the 1-d version of the Compute, there is a 1D, 2D overload of the compute. 

https://github.com/bertmaher/pytorch/blob/pytorch_fusion/torch/csrc/jit/tensorexpr/tests/expr_test.cpp#L55

**Pull request merged**